tools/Google.Cloud.Tools.GenerateSBOM/Program.cs

```diff
@@ -1,329 +1,329 @@
+        var spdx = SpdxFrom(nupkgPath);
+        var sbomPath = nupkgPath + "-sbom.spdx.json";
+        using (var file = File.Open(sbomPath, FileMode.OpenOrCreate,
+            FileAccess.Write))
+        {
+            JsonSerializer.Serialize(file, spdx, s_indented);
+        }
+        Console.WriteLine("Wrote " + sbomPath);
+    }
+
+    /// <summary>
+    /// Generates a SPDX SBOM from the given .nupkg file path.
+    /// </summary>
+    /// <param name="nupkgPath">Path to a .nupkg file.</param>
+    /// <returns>Filled out SPDX SBOM.</returns>
+    /// <exception cref="MultipleNuspecError">
+    ///     When the .nupkg contains multiple .nuspec files.
+    /// </exception>
+    /// <exception cref="MissingNuspecError">
+    ///     When the .nupkg contains no .nuspec files.
+    /// </exception>
+    private static Spdx SpdxFrom(string nupkgPath)
+    {
+        // Compute the SHA1 hash of the .nupkg file.
+        var sha1 = SHA1.Create();
+        byte[]? hash = null;
+        using (var file = File.OpenRead(nupkgPath))
+        {
+            hash = sha1.ComputeHash(file);
+        }
+
+        // Use the SHA1 hash to create a deterministic uuid.
+        var uuid = Uuid5FromHash(hash);
+
+        // Find the .nuspec metadata in the .nupkg and collect the fields we
+        // need.
+        NuspecMetadata? nuspecMetadata = null;
+        using (var zipFile = System.IO.Compression.ZipFile.OpenRead(nupkgPath))
+        {
+            foreach (var entry in zipFile.Entries)
+            {
+                if (Path.GetExtension(entry.Name).ToLower() == ".nuspec")
+                {
+                    if (nuspecMetadata == null)
+                    {
+                        nuspecMetadata = GetMetadata(entry.Open());
+                    }
+                    else
+                    {
+                        throw new MultipleNuspecError(nupkgPath);
+                    }
+                }
+            }
+        }
+        if (nuspecMetadata == null)
+        {
+            throw new MissingNuspecError(nupkgPath);
+        }
+
+        // Assemble a SPDX SBOM.
+        var fileName = Path.GetFileName(nupkgPath);
+        var spdx = new Spdx
+        {
+            name = nuspecMetadata.Id + "@" + nuspecMetadata.Version,
+            documentNamespace = "https://spdx.google/" + uuid,
+            creators = new List<string> { "Organization: Google LLC" },
+            created = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ",
+                System.Globalization.CultureInfo.InvariantCulture),
+            packages = new List<SpdxPackage> {
+                new() {
+                    name = fileName,
+                    SPDXID = "SPDXRef-Package-" + fileName + "-" + uuid,
+                    versionInfo = nuspecMetadata.Version,
+                    description = nuspecMetadata.Description,
+                    checksums = new List<SpdxChecksum> {
+                        new() {
+                            algorithm = "SHA1",
```

Comment:
Do we definitely want to use SHA-1 rather than SHA-256? My understanding is that SHA-1 should really not be used for anything new these days. Even if other SBOM creation tools use SHA-1, could we not use SHA-256?

Reply:
No checksum is required.  Since the code already calculates a SHA1 to build the UUID, I reckoned it was easy to record in the SBOM.